traffic_portal/app/src/common/modules/table/types/TableTypesController.js

```diff
@@ -33,15 +37,43 @@ var TableTypesController = function(types, $scope, $state, locationUtils) {
         $state.reload(); // reloads all the resolves for the view
     };
 
+    $scope.toggleVisibility = function(colName) {
+        const col = table.column(colName + ':name');
+        col.visible(!col.visible());
+        // hack alert: there is no api to set searchable on a column but if the column is visible, then it's searchable
+        table.context[0].aoColumns[col.index()].bSearchable = col.visible();
+        // redraw so the column's searchable value is taken into account
+        table.rows().invalidate().draw();
+    };
+
     angular.element(document).ready(function () {
-        $('#typesTable').dataTable({
-            "aLengthMenu": [[25, 50, 100, -1], [25, 50, 100, "All"]],
-            "iDisplayLength": 25,
-            "aaSorting": []
-        });
+        table = $('#typesTable').DataTable({
+                "lengthMenu": [[25, 50, 100, -1], [25, 50, 100, "All"]],
+                "iDisplayLength": 25,
+                "aaSorting": [],
+                "columns": [
+                    { "name": "name", "visible": true, "searchable": true },
+                    { "name": "description", "visible": true, "searchable": true },
+                    { "name": "lastUpdated", "visible": false, "searchable": false },
+                    { "name": "useInTable", "visible": true, "searchable": true }
+                ],
+                "colReorder": {
+                    realtime: false
+                },
+                "initComplete": function(settings, json) {
+                    // need to bind the show/hide column checkboxes to the saved visibility
+                    $scope.columns = JSON.parse($window.localStorage['DataTables_typesTable_/'])['columns'];
```

Comment:
The standard [Storage](https://developer.mozilla.org/en-US/docs/Web/API/Storage) interface doesn't appear to support indexing. It does work in Firefox, and I assume you tested it in Chrome, so maybe we don't care, but indexing seems to be non-standard and implementation-dependent. I'd recommend using `Storage.getItem()` instead.

This line also has an uncaught `SyntaxError` from that call to `JSON.parse`, which I guess is fine if it's somehow implicitly handled by the dataTables? Seems unlikely though.

You're also accessing a property of an object by indexing it with a string, which isn't exactly a problem in this case because of how the Storage API works, but it's a bad habit to get into because it forces type coercion on the indexed value.

Finally, you should check for the availability of the localStorage object ([MDN Guide](https://developer.mozilla.org/en-US/docs/Web/API/Web_Storage_API/Using_the_Web_Storage_API#Feature-detecting_localStorage)). This is because some browsers don't have localStorage (though I'm sure we don't support them), but also because sometimes it may just not be _available_ even if it is _supported_. As an example, Safari's private browsing mode gives you an empty object with 0 quota on every access. It's possible that dataTablesSort or whatever it's called does this already, but you should be sure; I'm not.

Reply:
ok, i'll see what i can clean up here. thanks